### ISSUE TITLE
[css-cascade-6] Add a `media()` function for `@import`

### ISSUE THREAD

**Issue by romainmenke** [w3c/csswg-drafts#10972]
https://github.com/w3c/csswg-drafts/issues/7348#issuecomment-2379746044

Media query lists can be partially valid.

This means that any `@import` that has both a media query list of at least two parts and any unknown condition will just swallow that unknown part as if it were an invalid media query.

https://codepen.io/romainmenke/pen/QWebVmp

```css
@import url("foo.css") screen, screen does-not-exist();
@import url("foo.css")  does-not-exist() screen, screen;
```

In either case `screen does-not-exist()` forms an invalid media query and the media query list will always be a singular `screen`.

Ideally we want unknown parts to invalidate the entire `@import`, not the first or last media query in a media query list.

------

### Proposal

Should we have a `media()` function to mirror `scope()`, `supports()`, ... ?
Wrapping it in a function makes it unambiguous what is part of a media query and what is not.

`@import` statements with both a `media()` function and a bare media query list should not be allowed. Bare media queries could even be marked as a "legacy syntax" and authors could be advised not to use it.

This function itself would suffer from the same problem it is trying to address but this would be a short term pain.

Because it is purely syntactic sugar we can transpile with tooling and authors could adopt this quickly.

-----

### Workaround

Any media query list can be split into multiple import statements:

```css
@import url("foo.css") screen;
@import url("foo.css") screen does-not-exist();

/* two imports for `screen` don't make sense in practice, but imagine something more creative and useful */
```

I think this is also fine and we can have tooling to help authors (lint rules, ...)
Drawback is that it isn't obvious that you sometimes need to do this and authors shouldn't need to use tool.

Even less obvious when you can clean this up.
At some point all end users of a project will support the newly added syntax and the statements can be joined again.

----

I prefer a new `media()` function as this doesn't have any sharp edges and gives a nice parallel between the other import conditions.

**Comment by andruud** [comment]
+1 to this.

> `@import` statements with both a media() function and a bare media query list should not be allowed. 

Additionally, I assume we'd probably want to not allow mixing of "new stuff" (just `scope()`, at the moment) and bare MQ.

Without `media()`, we'll _at least_ need to say that MQs always go last, although I think `media()` is a better solution.

cc @mirisuzanne

**Comment by tabatkins** [comment]
Yeah, I'll have to squint at the grammar to make sure that's enforceable, but if we could just branch it so an import *either* has a bare MQ *or* has all the new things (including `media()`), that would be ideal.

**Comment by cdoublev** [comment]
As noted in https://github.com/w3c/csswg-drafts/pull/11237#issuecomment-2505521838, defining `<media()> = media(<media-query-list>)` could be useful.

I do not know if [`if()`](https://drafts.csswg.org/css-values-5/#funcdef-if) intentionally uses `media(<media-query>)` but `color: if(media(unknown): green)` would be invalid, but not with `media(<media-query-list>)`, if I am not mistaken.

**Comment by romainmenke** [comment]
@tabatkins Is it correct that this resolution is to extend the syntax for `@import` so that these for example becomes valid syntax:
- `@import "foo.css" supports(width: 10px) or media(color);`
- `@import "foo.css" media(min-width: 200px) or media(color);`

If so, I am not sure if there is a use case for that 🤔 
Might be interesting but wasn't really what I had in mind :)

Does this allow anything that would not be possible by only adding a `media()` function that accepts a media query list?

-----

Is it also correct that media types would not be allowed in `media()`?
If so, does it then still solve the original issue?

If CSS authors still have a reason to write bare media query lists they will still encounter the original issue, that unrelated parts might get ignored as an invalid media query.

e.g. `@import "foo.css" media(color) print, something-we-havent-invented-yet;`

Or do we already know that the list of types is not going to change going forwards?
With `print` and `screen` being the only ones that actually have purpose.

**Comment by tabatkins** [comment]
> it correct that this resolution is to extend the syntax for @import so that these for example becomes valid syntax:

Yes, matching the syntax used for `if()` tests and the proposed `@when` generic-conditional rule. There's no reason to invent multiple slight variants of these syntaxes.

> Is it also correct that media types would not be allowed in media()?
> If so, does it then still solve the original issue?

Yes, media type is not allowed.

The resolution we ended up taking is that the trailing MQ *is* still allowed, so you can use a media type there if needed. Aside from that, tho, my and @frivoal's intention has always been to ensure that media features deprecate the need for media types, and at least in the case of `print` that should be sufficiently addressed by now.

> Or do we already know that the list of types is not going to change going forwards?
> With print and screen being the only ones that actually have purpose.

Correct, we consider media types to be a bad design and will not be extending the list further.

**Comment by romainmenke** [comment]
Sounds good!
Thank you 🙇

**Comment by andruud** [comment]
Combining the resolution in this issue with:

[`RESOLVED: allow full reordering of all the current conditions on @import (MQ, scope, ...)`](https://github.com/w3c/csswg-drafts/issues/7348#issuecomment-2377818033)

would allow e.g.:

```
@import "foo.css" supports(width: 10px) or media(color) layer(mylayer);
@import "foo.css" scope(.elem) supports(width: 10px) or media(color) layer(mylayer);
@import "foo.css" supports(width: 10px) or media(color) scope layer;
```

Are we happy with those aesthetics?

**Comment by romainmenke** [comment]
> Are we happy with those aesthetics?

Not personally.
I think it is confusing.
As if `scope layer` in the last example only applies to the `or media(color)` branch.

Requiring parenthesis is maybe better:

```
@import "foo.css" (supports(width: 10px) or media(color)) layer(mylayer);
@import "foo.css" scope(.elem) (supports(width: 10px) or media(color)) layer(mylayer);
@import "foo.css" (supports(width: 10px) or media(color)) scope layer;
```

**Comment by mirisuzanne** [comment]
Yeah, parenthesis or a slash? Something to act as a delimiter between the 'phrases' would be helpful.

**Comment by romainmenke** [comment]
I think a slash would be neat.
Makes it visually clear where the separation is.
